File: 48-rotate-image/rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        """
        (0,0) (0,1) (0,2) ... (0,n-1)
        (1,0) (1,1) ...       (1,n-1)
        (2,0) (2,1) ...       (2,n-1)
        (3,0) (3,1) ...       (3,n-1)
        (4,0) (4,1) ...       (4,n-1)
        ...
        (n-1,0) (n-1,1) ...   (n-1,n-1)
        """
        n = len(matrix)
        for offset in range(0, n//2):
            for i in range(offset, n-1-offset):
                matrix[offset][i], matrix[n-1-i][offset], matrix[n-1-offset][n-1-i], matrix[i][n-1-offset] = matrix[n-1-i][offset], matrix[n-1-offset][n-1-i], matrix[i][n-1-offset], matrix[offset][i]

File: 48-rotate-image/rotate-image.ts
/**
 Do not return anything, modify matrix in-place instead.
 */
function rotate(matrix: number[][]): void {
    for (let i = 0; i < matrix.length - 1; i++) {
        for (let j = i + 1; j < matrix[i].length; j++) {
            let temp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = temp;
        }
    }
    const k = Math.floor(matrix.length / 2);
    for (let j = 0; j < k; j++) {
        for (let i = 0; i < matrix.length; i++) {
            let temp = matrix[i][j];
            matrix[i][j] = matrix[i][matrix[j].length - 1 - j];
            matrix[i][matrix[j].length - 1 - j] = temp;
        }
    }
};